### PR TITLE
Implement Delta Snapshot Weight Synchronization & Vectorized Host-CPU Diffing Engine

### PR DESCRIPTION
**Pull request by droot**
## 🎯 High-Level Objective

  When training large Full Fine-Tuning (FFT) / Reinforcement Learning models (e.g.,  Qwen/Qwen3-8B ,  16.38 GB weights ) on distributed GPU clusters, transmitting full weight checkpoints across network storage on every RL rollout step creates massive I/O bottlenecks and temporary GPU VRAM spikes during checkpoint
  serialization.

  This PR introduces two foundational, composable synchronization engines to Open-RL that decouple parameter extraction from GPU memory and eliminate network I/O saturation:

  1. Delta Snapshot Weight Synchronization Engine ( strategy="delta" ): Extracted sparse delta checkpoints containing only modified/learned parameters (~2.0% changed elements per step) instead of full 16.38 GB snapshots.
  2. Multi-Threaded Vectorized Host-CPU Diffing Engine ( diffing_device="cpu" ): Offloads parameter comparisons ( cur_cpu.ne(prev_cpu) ) from GPU VRAM directly to Host DDR5 RAM across multi-core AVX-512 SIMD registers.
  ──────
  ## 🛠️ Core Architectural Changes

  ### 1. Delta Snapshot Weight Synchronization Engine ( strategy="delta" )

  • Core Engine ( src/training/fft_trainer_worker.py ):
      • Implemented  save_state_delta()  to compute sparse tensor differences against the previous step's shadow weight buffer.
      • Extracted sparse deltas ( ~330 MB on Qwen3-8B ) instead of full weights ( 16.38 GB ), reducing network storage write bandwidth requirements by ~98%.
  • Decoupled Synchronization from Tuning Mode ( Commit c06ff49 ):
      • Allowed  weight_sync_strategy  ( "full" | "delta" ) to be selected independently of fine-tuning mode ( LoRA  vs.  Full Fine-Tuning ).

  ### 2. Multi-Threaded Vectorized Host-CPU Diffing ( diffing_device="cpu" )

  • Zero-PCIe/Zero-VRAM Extraction ( Commit b668e37 ,  e6fa4fd ):
      • Added  diffing_device  selection ( "cpu" | "gpu" | "benchmark" ) to  FFTTrainingWorker.save_state_delta .
      • Under  "cpu" , the worker compares current and historical weights on Host DDR5 RAM across 26 vCPUs ( vectorized AVX-512 SIMD comparison ), eliminating 16.38 GB of H2D PCIe shadow copying and 16.71 GB of temporary GPU VRAM overhead per step.
      • Set  diffing_device="cpu"  as the global fallback default ( Release 0.1.55+ ).
  • Deterministic Unit Testing ( tests/test_diffing_backends.py ):
      • Created an exhaustive unit test suite verifying exact numerical sparsity, metadata density, and tensor shape parity across  "cpu" ,  "gpu" , and  "benchmark"  modes ( 81 unit tests passing ).

  ### 3. Per-Job Header Plumbing & Pod Isolation

  • HTTP Header Plumbing ( src/server/gateway.py ,  sitecustomize.py ,  tinker_utils.py ):
      • Added dynamic forwarding of  OPEN_RL_WEIGHT_SYNC_STRATEGY  ( X-Open-RL-Weight-Sync-Strategy ) and  OPEN_RL_DIFFING_DEVICE  ( X-Open-RL-Diffing-Device ) HTTP headers from client scripts through the API Gateway down to each request.
  • Per-Tenant Kubernetes Pod Isolation ( src/server/k8s_worker_manager.py ):
      • Updated  render_pod()  and  _prepare_pod_spec()  to inject specific environment variables ( OPEN_RL_WEIGHT_SYNC_STRATEGY ,  OPEN_RL_DIFFING_DEVICE ) directly into individual  open-rl-trainer-<job_id>  worker pod manifests, preventing crosstalk during concurrent multi-tenant execution.

  ──────
  ## 🧪 E2E Test Harness Scenarios ( scripts/run_training_e2e.py )

  Added three dedicated integration scenarios verified on Kubernetes ( Qwen/Qwen3-8B , 192 batch size per step):

  1.  fft-gsm8k-rl  (Delta Parity Check): Verified single-tenant training under delta sync vs full sync across 20–30 steps.
  2.  fft-gsm8k-rl-x2-compare : Concurrently interleaves Job A ( Full Sync ) vs. Job B ( Delta Sync ) on the same GPU via the Accelerator Time-Slicer.
  3.  fft-gsm8k-rl-x2-diffing-compare : Concurrently interleaves Job A ( GPU Diffing - 'gpu' ) vs. Job B ( Host-CPU Diffing - 'cpu' ) under delta sync for 15 steps.
  ──────
  ## 📊 Empirical Benchmarks & Verification Highlights

  1. Mathematical & Convergence Parity:
      • Across all head-to-head campaigns ( single-tenant 30 steps  &  time-sliced 15 steps ), delta synchronization and Host-CPU diffing achieved exact reward parity and accuracy trajectories ( converging from ~8% up to ~98% accuracy on GSM8K ).
  2. Single-Tenant Checkpoint Speedup ( -33.7% ):
      • On H100 single-tenant runs ( Qwen3-8B ), switching from GPU diffing ( 103.1s ) to Host-CPU diffing ( 68.4s ) reduced steady-state  time/save_checkpoint  latency by 34.7 seconds (-33.7%).
  3. Multi-Tenant Memory Safety & Collocation Density:
      • During concurrent multi-tenant execution ( fft-gsm8k-rl-x2-diffing-compare ), while Job A ( GPU diffing ) consumes an extra 16.71 GB of temporary VRAM to execute  .ne().nonzero()  on GPU tensors, Job B ( Host-CPU diffing ) consumes 0 additional VRAM, preventing CUDA out-of-memory crashes when multiple
      large-model workers checkpoint concurrently.

### Qwen/Qwen3-8B ( 16.38 GB ) Full vs. Delta Weight Sync Benchmark ( v0.1.75 )

   Step | Full Acc ( Job A ) | Delta Acc ( Job B ) | Full Reward ( Job A… | Delta Reward ( Job B… | Delta Weight Chang… | Full Save ( Job A ) | Delta Save ( Job B ) | Full Total ( Job A ) | Delta Total ( Job B )
  ------|--------------------|---------------------|----------------------|-----------------------|---------------------|---------------------|----------------------|----------------------|-----------------------
    0   |       11.98%       |       10.94%        |       +0.0292        |        +0.0156        |       12.90%        |       65.94s        |        41.16s        |       410.05s        |        498.00s
    1   |       38.54%       |       35.94%        |       +0.3203        |        +0.2859        |        9.24%        |       81.06s        |        26.93s        |       368.83s        |        328.40s
    2   |       51.56%       |       60.42%        |       +0.4641        |        +0.5375        |        7.37%        |       99.83s        |        35.50s        |       374.84s        |        364.69s
    3   |       49.48%       |       59.38%        |       +0.4307        |        +0.5214        |        6.54%        |       82.30s        |        20.22s        |       357.25s        |        247.47s
    4   |       59.38%       |       66.67%        |       +0.5365        |        +0.6063        |        5.89%        |       75.19s        |        19.06s        |       399.72s        |        113.46s
    5   |       92.19%       |       90.10%        |       +0.9073        |        +0.8760        |        5.36%        |       73.25s        |        10.21s        |       349.30s        |        280.58s
    6   |       88.54%       |       90.62%        |       +0.8609        |        +0.8776        |        4.96%        |       78.59s        |        12.19s        |       337.29s        |        114.02s
    7   |         —          |       87.50%        |          —           |        +0.8484        |        4.71%        |          —          |        12.19s        |          —           |        237.03s
    8   |         —          |       95.31%        |          —           |        +0.9438        |        4.35%        |          —          |        19.84s        |          —           |        121.76s
    9   |         —          |       98.96%        |          —           |        +0.9703        |        4.10%        |          —          |        19.86s        |          —           |        329.11s
    10  |         —          |       95.31%        |          —           |        +0.9448        |        3.86%        |          —          |        9.59s         |          —           |        248.50s

### CODE REVIEW

File: src/server/vllm_sampler.py [hunk 1]
```diff
@@ -60,6 +64,138 @@ def is_fft_enabled() -> bool:
   time_slicer = time_slicer_client_from_env()
 
 
+def patch_vllm_worker_for_delta_sync():
```
Comment (ShubyM) [inline src/server/vllm_sampler.py]: This is not needed if we have the implementation that uses VLLM hook

File: src/training/fft_trainer_worker.py [hunk 2]
```diff
@@ -157,6 +171,109 @@ def save_state(self, model_id: str, state_path: str, include_optimizer: bool = F
...
+      diffing_device = diffing_device.lower()
+
+    os.makedirs(state_path, exist_ok=True)
+    was_offloaded = self._prepare_for_save()
+    delta_tensors: dict[str, torch.Tensor] = {}
+    total_changed = 0
+    total_elements = 0
+
+    try:
+      if hasattr(self, "_latest_delta_tensors") and self._latest_delta_tensors:
```
Comment (ShubyM) [inline src/training/fft_trainer_worker.py]: Small thing but the models love to be overly defensive and use hasattr and getattr instead of just self.property

File: src/training/fft_trainer_worker.py [hunk 3]
```diff
@@ -262,6 +379,29 @@ def sleep(self) -> None:
...
+    # 1-CPU-Copy (15.26 GB) Offload-Time Sparse Delta Diffing:
+    # Compare live GPU W_{t+1} against existing CPU _param_shadow (holding W_t) before overwriting
+    if hasattr(self, "_latest_delta_tensors"):
+      self._latest_delta_tensors.clear()
+    else:
+      self._latest_delta_tensors = {}
+    self._latest_total_changed = 0
+    self._latest_total_elements = 0
+
+    for name, param in self.model.named_parameters():
```
Comment (ShubyM) [inline src/training/fft_trainer_worker.py]: Does this unconditionally do a GPU diff? I think if I'm following correctly we have two choices for where diffs occur cpu or gpu. We always store the previous set of weights in CPU memory and the current set of weights in GPU memory.  In this implementation, as soon as sleep is fired we transfer the previous CPU weights to the GPU and always compute the diff there, then we transfer the current set of weights back to CPU. We could make this respect the diff device selection and transfer the current set of weights back to CPU memory and then compute the diff there.

File: src/server/vllm_sampler.py (hunk 1 above)
Comment (droot) [reply src/server/vllm_sampler.py]: Good catch. Removed this. Though we ended up using vllm's non-public methods to initializing the base model to implement those hooks. We also had to patch some stuff in vllm as well.

This might need some work when we upgrade to latest vllm (which I guess we need for the Legal agent fine-tuning).

File: src/training/fft_trainer_worker.py (hunk 2 above)
Comment (droot) [reply src/training/fft_trainer_worker.py]: Addressed this. Also added tip in agents.md.

File: src/training/fft_trainer_worker.py (hunk 3 above)
Comment (droot) [reply src/training/fft_trainer_worker.py]: Thanks for catching this. I have cleaned this up. Now we do the delta computation in `optim_step` because that's where weights are updated and decoupled it from sleep. While reviewing this, I also found that we were keeping an extra copy of previous weights (in addition to offloaded state in cpu) just for delta computation, now we just reuse the offloaded state for delta computation.

**Comment by droot** [comment]
@ShubyM the PR is now ready for final review.

File: src/server/vllm_sampler.py (hunk 1 above)
Comment (droot) [reply src/server/vllm_sampler.py]: I ended up upgrading vllm to latest v0.25.1. It did make the docker images bigger. I am wondering with this upgrade, do we still need the patch for Gemma.

**Comment by droot** [comment]
> @ShubyM the PR is now ready for final review.

hold for a bit. decoupling the diffing from sleep() and using the offloaded state had lots of inefficiency, so need to take another look.

**Comment by droot** [comment]
> > @ShubyM the PR is now ready for final review.
> 
> hold for a bit. decoupling the diffing from sleep() and using the offloaded state had lots of inefficiency, so need to take another look.

ok, so it is ready for final review. I will do a separate PR for migrating upstream/fft to `v0.25.1`.

I have also done some clean up in the API layer and request handling. Removed the diffing_device concept completely. Now we do diffing on GPU. There is scope for further improving the performance of diffing, but it is in good place now.

I also experimented with multi-agent workflow where I will ask one agent to come up with the design (you can see two examples in the PR) and then I will ask another agent to review the design and cross check and refine it. and then I will go back or launch a fresh agent to implement the changes as per the design and ask the agent to update the design as they make the changes. I will keep the diff open in my IDE and keep on further steering but keeping the design up to date. I think I had a reasonable success with this, but I was completely in the loop with all the three agents.

Pl. take a look.

**Review by ShubyM — APPROVED** [review]
